src/app/modules/form/admin-form/admin-form.controller.ts

```diff
@@ -1730,6 +1733,106 @@ export const handleReorderFormField = [
   _handleReorderFormField,
 ] as RequestHandler[]
 
+/**
+ * NOTE: Exported for testing.
+ * Private handler for PUT /forms/:formId/logic/:logicId
+ * @precondition Must be preceded by request validation
+ * @security session
+ *
+ * @returns 200 with success message and updated logic object when successfully updated
+ * @returns 403 when user does not have permissions to update logic
+ * @returns 404 when form cannot be found
+ * @returns 422 when user in session cannot be retrieved from the database
+ * @returns 500 when database error occurs
+ */
+export const _handleUpdateLogic: RequestHandler<
+  { formId: string; logicId: string },
+  LogicDto | ErrorDto,
+  LogicUpdateDto
+> = (req, res) => {
+  const { formId, logicId } = req.params
+  const { updatedLogic } = req.body
+  const sessionUserId = (req.session as Express.AuthedSession).user._id
+
+  // Step 1: Retrieve currently logged in user.
+  return (
+    UserService.getPopulatedUserById(sessionUserId)
+      .andThen((user) =>
+        // Step 2: Retrieve form with write permission check.
+        AuthService.getFormAfterPermissionChecks({
+          user,
+          formId,
+          level: PermissionLevel.Write,
+        }),
+      )
+      // Step 3: Update form logic
+      .andThen((retrievedForm) =>
+        AdminFormService.updateFormLogic(retrievedForm, logicId, updatedLogic),
+      )
+      .map((updatedLogic) => res.status(StatusCodes.OK).json(updatedLogic))
+      .mapErr((error) => {
+        logger.error({
+          message: 'Error occurred when updating form logic',
+          meta: {
+            action: 'handleUpdateLogic',
+            ...createReqMeta(req),
+            userId: sessionUserId,
+            formId,
+            logicId,
+            updatedLogic,
+          },
+          error,
+        })
+        const { errorMessage, statusCode } = mapRouteError(error)
+        return res.status(statusCode).json({ message: errorMessage })
+      })
+  )
+}
+
+/**
+ * Handler for PUT /forms/:formId/logic/:logicId
+ */
+export const handleUpdateLogic = [
+  celebrate(
+    {
+      [Segments.BODY]: Joi.object({
+        // Ensures given logic is same as accessed logic
+        updatedLogic: {
```

Comment:
is it necessary to have `updatedLogic` as an intermediate key? if not, would it be better to remove `LogicUpdateDto` and just have `LogicDto` for both the request and response?

Reply:
edited